File: cast-extractor/src/main/java/com/libreshockwave/tools/scanning/FileProcessor.java
package com.libreshockwave.tools.scanning;

import com.libreshockwave.DirectorFile;
import com.libreshockwave.cast.BitmapInfo;
import com.libreshockwave.cast.FilmLoopInfo;
import com.libreshockwave.cast.MemberType;
import com.libreshockwave.cast.ShapeInfo;
import com.libreshockwave.chunks.*;
import com.libreshockwave.format.ScriptFormatUtils;
import com.libreshockwave.tools.model.CastMemberInfo;
import com.libreshockwave.tools.model.FileNode;

import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Processes Director files and extracts member information.
 */
public class FileProcessor {

    private final Map<String, DirectorFile> loadedFiles;
    private boolean lazyLoading = false;

    public FileProcessor(Map<String, DirectorFile> loadedFiles) {
        this.loadedFiles = loadedFiles;
    }

    /**
     * Enable lazy loading mode - don't keep DirectorFiles in memory during scan.
     * Files will be reloaded on demand when needed for preview/export.
     * This significantly reduces memory usage for large scans.
     */
    public void setLazyLoading(boolean lazyLoading) {
        this.lazyLoading = lazyLoading;
    }

    public boolean isLazyLoading() {
        return lazyLoading;
    }

    /**
     * Process a single file and return a FileNode, or null if no valid members.
     * This method is thread-safe.
     */
    public FileNode processFile(Path file) {
        try {
            DirectorFile dirFile = DirectorFile.load(file);
            List<CastMemberInfo> members = new ArrayList<>();

            // Get all cast members
            for (CastMemberChunk member : dirFile.getCastMembers()) {
                if (member.memberType() == MemberType.NULL) {
                    continue; // Skip null members
                }

                String name = member.name();
                if (name == null || name.isEmpty()) {
                    name = "Unnamed #" + member.id();
                }

                String details = buildMemberDetails(dirFile, member);

                members.add(new CastMemberInfo(
                        member.id(), name, member, member.memberType(), details
                ));
            }

            if (!members.isEmpty()) {
                // In lazy loading mode, don't keep the file in memory
                // It will be reloaded on demand when needed
                if (!lazyLoading) {
                    loadedFiles.put(file.toString(), dirFile);
                }
                return new FileNode(file.toString(), file.getFileName().toString(), members);
            }
        } catch (Exception ignored) {
            // Silently ignore files that fail to parse
        }
        return null;
    }

    private String buildMemberDetails(DirectorFile dirFile, CastMemberChunk member) {
        MemberType type = member.memberType();
        String details = "";

        // Parse type-specific details
        if (type == MemberType.BITMAP && member.specificData().length > 0) {
            try {
                BitmapInfo info = BitmapInfo.parse(member.specificData());
                details = String.format("%dx%d, %d-bit", info.width(), info.height(), info.bitDepth());
            } catch (Exception ignored) {}
        } else if (type == MemberType.SHAPE && member.specificData().length > 0) {
            try {
                ShapeInfo info = ShapeInfo.parse(member.specificData());
                details = String.format("%s %dx%d", info.shapeType(), info.width(), info.height());
            } catch (Exception ignored) {}
        } else if (type == MemberType.FILM_LOOP && member.specificData().length > 0) {
            try {
                FilmLoopInfo info = FilmLoopInfo.parse(member.specificData());
                details = String.format("%dx%d", info.width(), info.height());
            } catch (Exception ignored) {}
        } else if (type == MemberType.SCRIPT) {
            details = buildScriptDetails(dirFile, member);
        } else if (type == MemberType.SOUND) {
            details = buildSoundDetails(dirFile, member);
        } else if (type == MemberType.PALETTE) {
            details = buildPaletteDetails(dirFile, member);
        } else if (type == MemberType.TEXT || type == MemberType.BUTTON) {
            details = buildTextDetails(dirFile, member);
        }

        return details;
    }

    private String buildScriptDetails(DirectorFile dirFile, CastMemberChunk member) {
        ScriptChunk script = MemberResolver.findScriptForMember(dirFile, member);
        if (script != null) {
            ScriptNamesChunk scriptNames = dirFile.getScriptNames();
            String scriptTypeName = ScriptFormatUtils.getScriptTypeName(script.getScriptType());

            // Build handler list summary
            List<String> handlerNames = new ArrayList<>();
            if (scriptNames != null) {
                for (ScriptChunk.Handler h : script.handlers()) {
                    String hName = ScriptFormatUtils.resolveName(scriptNames, h.nameId());
                    if (!hName.startsWith("<")) {
                        handlerNames.add(hName);
                    }
                }
            }

            // Format: "Movie Script" or "Parent Script" with handler names
            if (!handlerNames.isEmpty()) {
                String handlers = handlerNames.size() <= 3 ?
                        String.join(", ", handlerNames) :
                        handlerNames.get(0) + ", " + handlerNames.get(1) + "... +" + (handlerNames.size() - 2);
                return String.format("%s [%s]", scriptTypeName, handlers);
            } else {
                return scriptTypeName;
            }
        }
        return "";
    }

    private String buildSoundDetails(DirectorFile dirFile, CastMemberChunk member) {
        SoundChunk soundChunk = MemberResolver.findSoundForMember(dirFile, member);
        if (soundChunk != null) {
            String codec = soundChunk.isMp3() ? "MP3" : "PCM";
            double duration = soundChunk.durationSeconds();
            return String.format("%s, %dHz, %.1fs", codec, soundChunk.sampleRate(), duration);
        }
        return "sound data";
    }

    private String buildPaletteDetails(DirectorFile dirFile, CastMemberChunk member) {
        PaletteChunk paletteChunk = MemberResolver.findPaletteForMember(dirFile, member);
        if (paletteChunk != null) {
            return paletteChunk.colors().length + " colors";
        }
        return "palette";
    }

    private String buildTextDetails(DirectorFile dirFile, CastMemberChunk member) {
        TextChunk textChunk = MemberResolver.findTextForMember(dirFile, member);
        if (textChunk != null) {
            String text = ScriptFormatUtils.normalizeLineEndings(textChunk.text());
            text = ScriptFormatUtils.truncate(text, 50);
            return "\"" + text + "\"";
        }
        return "";
    }
}

File: cpp/src/editor/scanning/FileProcessor.cpp
#include "libreshockwave/editor/scanning/FileProcessor.hpp"

#include <exception>
#include <iomanip>
#include <memory>
#include <sstream>
#include <string>
#include <utility>
#include <vector>

#include "libreshockwave/DirectorFile.hpp"
#include "libreshockwave/cast/BitmapInfo.hpp"
#include "libreshockwave/cast/FilmLoopInfo.hpp"
#include "libreshockwave/cast/MemberType.hpp"
#include "libreshockwave/cast/ShapeInfo.hpp"
#include "libreshockwave/chunks/CastMemberChunk.hpp"
#include "libreshockwave/chunks/PaletteChunk.hpp"
#include "libreshockwave/chunks/ScriptNamesChunk.hpp"
#include "libreshockwave/chunks/SoundChunk.hpp"
#include "libreshockwave/chunks/TextChunk.hpp"
#include "libreshockwave/editor/scanning/MemberResolver.hpp"
#include "libreshockwave/format/ScriptFormatUtils.hpp"

namespace libreshockwave::editor::scanning {
namespace {

std::string shapeTypeName(cast::ShapeType shapeType) {
    switch (shapeType) {
        case cast::ShapeType::Rect:
            return "RECT";
        case cast::ShapeType::OvalRect:
            return "OVAL_RECT";
        case cast::ShapeType::Oval:
            return "OVAL";
        case cast::ShapeType::Line:
            return "LINE";
        case cast::ShapeType::Unknown:
            return "UNKNOWN";
    }
    return "UNKNOWN";
}

std::string formatOneDecimal(double value) {
    std::ostringstream out;
    out << std::fixed << std::setprecision(1) << value;
    return out.str();
}

} // namespace

std::vector<model::CastMemberInfo> FileProcessor::processMembers(DirectorFile& dirFile) const {
    std::vector<model::CastMemberInfo> members;
    for (const auto& member : dirFile.castMembers()) {
        if (member == nullptr || member->memberType() == cast::MemberType::Null) {
            continue;
        }

        auto name = member->name();
        if (name.empty()) {
            name = "Unnamed #" + std::to_string(member->id().value());
        }
        members.push_back(model::CastMemberInfo{
            member->id().value(),
            std::move(name),
            member,
            member->memberType(),
            buildMemberDetails(dirFile, member)
        });
    }
    return members;
}

std::string FileProcessor::buildMemberDetails(DirectorFile& dirFile,
                                              const std::shared_ptr<chunks::CastMemberChunk>& member) const {
    if (member == nullptr) {
        return "";
    }

    const auto type = member->memberType();
    const auto& specificData = member->specificData();
    if (type == cast::MemberType::Bitmap && !specificData.empty()) {
        try {
            const auto info = cast::BitmapInfo::parse(specificData);
            return std::to_string(info.width) + "x" + std::to_string(info.height) + ", " +
                   std::to_string(info.bitDepth) + "-bit";
        } catch (const std::exception&) {
            return "";
        }
    }
    if (type == cast::MemberType::Shape && !specificData.empty()) {
        try {
            const auto info = cast::ShapeInfo::parse(specificData);
            return shapeTypeName(info.shapeType) + " " + std::to_string(info.width) + "x" +
                   std::to_string(info.height);
        } catch (const std::exception&) {
            return "";
        }
    }
    if (type == cast::MemberType::FilmLoop && !specificData.empty()) {
        try {
            const auto info = cast::FilmLoopInfo::parse(specificData);
            return std::to_string(info.width()) + "x" + std::to_string(info.height());
        } catch (const std::exception&) {
            return "";
        }
    }
    if (type == cast::MemberType::Script) {
        return buildScriptDetails(dirFile, member);
    }
    if (type == cast::MemberType::Sound) {
        return buildSoundDetails(dirFile, member);
    }
    if (type == cast::MemberType::Palette) {
        return buildPaletteDetails(dirFile, member);
    }
    if (type == cast::MemberType::Text || type == cast::MemberType::Button) {
        return buildTextDetails(dirFile, member);
    }
    return "";
}

std::string FileProcessor::buildScriptDetails(DirectorFile& dirFile,
                                              const std::shared_ptr<chunks::CastMemberChunk>& member) const {
    const auto script = MemberResolver::findScriptForMember(dirFile, member);
    if (script == nullptr) {
        return "";
    }

    const auto scriptNames = dirFile.scriptNames();
    const auto scriptTypeName = format::getScriptTypeName(script->scriptType());
    std::vector<std::string> handlerNames;
    if (scriptNames != nullptr) {
        handlerNames.reserve(script->handlers().size());
        for (const auto& handler : script->handlers()) {
            auto handlerName = format::resolveName(scriptNames.get(), handler.nameId);
            if (!handlerName.starts_with("<")) {
                handlerNames.push_back(std::move(handlerName));
            }
        }
    }

    if (handlerNames.empty()) {
        return scriptTypeName;
    }

    std::string handlers;
    if (handlerNames.size() <= 3) {
        handlers = handlerNames.front();
        for (std::size_t index = 1; index < handlerNames.size(); ++index) {
            handlers += ", " + handlerNames[index];
        }
    } else {
        handlers = handlerNames[0] + ", " + handlerNames[1] + "... +" +
                   std::to_string(handlerNames.size() - 2);
    }
    return scriptTypeName + " [" + handlers + "]";
}

std::string FileProcessor::buildSoundDetails(DirectorFile& dirFile,
                                             const std::shared_ptr<chunks::CastMemberChunk>& member) const {
    const auto sound = MemberResolver::findSoundForMember(dirFile, member);
    if (sound == nullptr) {
        return "sound data";
    }
    const std::string codec = sound->isMp3() ? "MP3" : "PCM";
    return codec + ", " + std::to_string(sound->sampleRate()) + "Hz, " +
           formatOneDecimal(sound->durationSeconds()) + "s";
}

std::string FileProcessor::buildPaletteDetails(DirectorFile& dirFile,
                                               const std::shared_ptr<chunks::CastMemberChunk>& member) const {
    const auto palette = MemberResolver::findPaletteForMember(dirFile, member);
    if (palette != nullptr) {
        return std::to_string(palette->colors().size()) + " colors";
    }
    return "palette";
}

std::string FileProcessor::buildTextDetails(DirectorFile& dirFile,
                                            const std::shared_ptr<chunks::CastMemberChunk>& member) const {
    const auto text = MemberResolver::findTextForMember(dirFile, member);
    if (text == nullptr) {
        return "";
    }
    auto normalized = format::normalizeLineEndings(text->text());
    normalized = format::truncate(normalized, 50);
    return "\"" + normalized + "\"";
}

} // namespace libreshockwave::editor::scanning
